tests/core/featurizers/test_tracker_featurizer.py

```diff
@@ -49,48 +64,2238 @@ def test_convert_labels_to_ids(domain: Domain):
+    )
+
+    expected_states = [
+        [
+            {},
+            {
+                USER: {INTENT: "greet"},
+                PREVIOUS_ACTION: {ACTION_NAME: ACTION_LISTEN_NAME},
+            },
+            {USER: {INTENT: "greet"}, PREVIOUS_ACTION: {ACTION_NAME: "utter_greet"},},
+            {
+                USER: {INTENT: "mood_great"},
+                PREVIOUS_ACTION: {ACTION_NAME: ACTION_LISTEN_NAME},
+            },
+            {
+                USER: {INTENT: "mood_great"},
+                PREVIOUS_ACTION: {ACTION_NAME: "utter_happy"},
+            },
+        ]
+    ]
+
+    if max_history is not None:
+        expected_states = [x[-max_history:] for x in expected_states]
+
+    assert actual_states is not None
+    assert len(actual_states) == len(expected_states)
+
+    for actual, expected in zip(actual_states, expected_states):
+        assert actual == expected
+
+
+@pytest.mark.parametrize("max_history", [None, 3])
+def test_prediction_states_keeps_action_intent_unlikely_with_intent_max_history_tracker_featurizer(
+    moodbot_tracker: DialogueStateTracker,
+    moodbot_domain: Domain,
+    max_history: Optional[int],
+):
+
+    state_featurizer = IntentTokenizerSingleStateFeaturizer()
+    tracker_featurizer = IntentMaxHistoryTrackerFeaturizer(
+        state_featurizer, max_history=max_history
+    )
+
+    tracker = DialogueStateTracker.from_events(
+        "default",
+        [
+            ActionExecuted(ACTION_LISTEN_NAME),
+            user_uttered("greet"),
+            ActionExecuted(ACTION_UNLIKELY_INTENT_NAME),
+            ActionExecuted("utter_greet"),
+            ActionExecuted(ACTION_LISTEN_NAME),
+            user_uttered("mood_great"),
+            ActionExecuted(ACTION_UNLIKELY_INTENT_NAME),
+            ActionExecuted("utter_happy"),
+            ActionExecuted(ACTION_LISTEN_NAME),
+            user_uttered("goodbye"),
+        ],
+        domain=moodbot_domain,
+    )
+
+    actual_states = tracker_featurizer.prediction_states([tracker], moodbot_domain)
+
+    expected_states = [
+        [
+            {},
+            {
+                USER: {INTENT: "greet"},
+                PREVIOUS_ACTION: {ACTION_NAME: ACTION_LISTEN_NAME},
+            },
+            {
+                USER: {INTENT: "greet"},
+                PREVIOUS_ACTION: {ACTION_NAME: ACTION_UNLIKELY_INTENT_NAME},
+            },
+            {USER: {INTENT: "greet"}, PREVIOUS_ACTION: {ACTION_NAME: "utter_greet"},},
+            {
+                USER: {INTENT: "mood_great"},
+                PREVIOUS_ACTION: {ACTION_NAME: ACTION_LISTEN_NAME},
+            },
+            {
+                USER: {INTENT: "mood_great"},
+                PREVIOUS_ACTION: {ACTION_NAME: ACTION_UNLIKELY_INTENT_NAME},
+            },
+            {
+                USER: {INTENT: "mood_great"},
+                PREVIOUS_ACTION: {ACTION_NAME: "utter_happy"},
+            },
+        ]
+    ]
```

Comment:
I'm confused why we don't flip intents and actions for intent ted. 
In this example you have: intent greet, followed by three possible actions. 
To train intent ted, dont you need the opposite? each action and the possible intents that can follow?

Reply:
[This comment](https://github.com/RasaHQ/rasa/pull/8570#issuecomment-849965353) also address this question.

Reply:
Thanks.